editor/src/components/canvas/dom-walker.ts

```diff
@@ -307,16 +311,72 @@ function useInvalidateInitCompleteOnMountCount(
 
   return [initCompleteRef.current, setInitComplete]
 }
+type ValueOrUpdater<S> = S | ((prevState: S) => S)
+// todo move to file
+export type SetValueCallback<S> = (
+  valueOrUpdater: ValueOrUpdater<S>,
+  doNotInvalidate?: 'do-not-invalidate',
+) => void
+
+function isSimpleValue<S>(valueOrUpdater: ValueOrUpdater<S>): valueOrUpdater is S {
+  return typeof valueOrUpdater !== 'function'
+}
+function useStateAsyncInvalidate<S>(
+  onInvalidate: () => void,
+  initialState: S,
+): [S, SetValueCallback<S>] {
+  const stateRef = React.useRef(initialState)
+  const setAndMarkInvalidated = React.useCallback(
+    (valueOrUpdater: ValueOrUpdater<S>, doNotInvalidate?: 'do-not-invalidate') => {
```

Comment:
Wait, the text `'do-not-invalidate' | undefined` is the type? I find this a bit confusing.

Reply:
you either omit this parameter, or you write `'no-not-invalidate'`. I typed it like this to avoid a magic `..., true)` on the call site

Reply:
I think if we want to avoid booleans it's much clearer to have a type of `'leave-alone' | 'invalidate'` than make it optional.

Reply:
+1 to that